### ISSUE TITLE
Joint Learning & Edge Splitting Issue

### ISSUE THREAD

**Issue by fffuuuming** [NTU-king-james/EllipticPlusPlus#2]
# Decription

Currently basic model doesn't perform well on test set, and the AE Loss doesn't improve during training, possibly due to the following reasons:
### Missused of RandomLinkSplit
- ```RandomLinkSplit```randomly selects a subset of existing edges as train/val/test
    - May not be a problem in our case, since ```num_val, num_test = 0```
- ```RandomLinkSplit``` samples negative edges randomly from non-existent edges
    - May sample negative edges between nodes that never existed at the same time (in different timestamp)
### Imbalanced labeled data during joint training
- ```z = encoder(x, edge_index```) is shared, so the gradient from the classifier also updates the encoder.
    - Most of the nodes are unlabeled, but z for all nodes is still computed.
    - Classifier is only trained on labeled data, but its gradients go through the entire encoder.
    - The encoder gets pulled in a direction to optimize classification on only **a small subset of nodes**, which may hurt the encoder’s ability to generalize for reconstruction of edge structure and cause overfitting to labeled nodes
### Possible Solutions
For splitting problem:
- Custom negative sampling restricted to nodes coexisting in the same time slice

For imbalance problem:
- **Balance data**
Keep only nodes with known labels and their neighbors, and only keep edges like ```unknown ↔ known```
    - may **destroy global graph structure**, which might hurt the GAE’s ability to generalize.
    - Comprise: May can keep ```Unknown–unknown```, if **≤ 2 hops from any known node**
- **Weighted joint loss**: ```loss = λ * loss_ae + (1 - λ) * loss_cls```
Can reduce λ over time
- **Two Stage Training**
Train GAE on link prediction -> get **z**, then fix **z** to train classifier on labeled nodes only, just as **inspection-L**

**Comment by fffuuuming** [comment]
Currently working on splitting task, since the performance isn't well even for train for AE Loss only
